experiments/targeting.go

```diff
@@ -0,0 +1,298 @@
+	}
+	for _, value := range n.values {
+		if candidateValue == value {
+			return true
+		}
+	}
+	return false
+}
+
+type NotNode struct {
+	child Targeting
+}
+
+func NewNotNode(inputNodes map[string]interface{}) (*NotNode, error) {
+	if len(inputNodes) != 1 {
+		return nil, TargetingNodeError("NotNode expects exactly one field")
+	}
+	node, err := parseNode(inputNodes)
+	if err != nil {
+		return nil, err
+	}
+	return &NotNode{
+		child: node,
+	}, nil
+}
+
+func (n *NotNode) Evaluate(inputs map[string]interface{}) bool {
+	return !n.child.Evaluate(inputs)
+}
+
+type OverrideNode struct {
+	ReturnValue bool
+}
+
+func NewOverrideNode(inputNode interface{}) *OverrideNode {
+	returnValue, ok := inputNode.(bool)
+	if !ok {
+		returnValue = false
+	}
+	return &OverrideNode{
+		ReturnValue: returnValue,
+	}
+}
+
+func (n *OverrideNode) Evaluate(inputs map[string]interface{}) bool {
+	return n.ReturnValue
+}
+
+type ComparisonNode struct {
+	field string
+	value interface{}
+	less  Less
+}
+
+func NewComparisonNode(inputs map[string]interface{}, less Less) (*ComparisonNode, error) {
+	if len(inputs) != 2 {
+		return nil, TargetingNodeError("ComparisonNode expects exactly two fields")
+	}
+	field, ok := inputs["field"]
+	if !ok {
+		return nil, TargetingNodeError("ComparisonNode expects input key 'field'.")
+	}
+	value, ok := inputs["value"]
+	if !ok {
+		return nil, TargetingNodeError("ComparisonNode expects input key 'value'.")
+	}
+	return &ComparisonNode{
+		field: field.(string),
+		value: value,
+		less:  less,
+	}, nil
+}
+
+func (n *ComparisonNode) Evaluate(inputs map[string]interface{}) bool {
+	candidateValue := inputs[n.field]
+	if n.value == nil {
+		return false
+	}
+	value, err := (n.value.(json.Number)).Float64()
+	if err != nil {
+		return false
+	}
+	switch cv := candidateValue.(type) {
+	case int:
+		return n.less(float64(cv), value)
+	case float64:
+		n.less(cv, value)
+	}
+	return false
+}
+
+type Targeting interface {
+	Evaluate(inputs map[string]interface{}) bool
+}
+
+func NewTargeting(targetingConfig []byte) (Targeting, error) {
+	// TODO handle empty input
+	var config map[string]interface{}
+	decoder := json.NewDecoder(bytes.NewBuffer(targetingConfig))
+	decoder.UseNumber()
+	err := decoder.Decode(&config)
+	if err != nil {
+		return nil, err
+	}
+	if len(config) != 1 {
+		return nil, TargetingNodeError("call to create targeting tree expects single input key")
+	}
+	return parseNodes(config)
+}
+
+func parseNodes(config map[string]interface{}) (Targeting, error) {
+	key := firstKeyStr(config)
+	return mapOperatorNode(strings.ToLower(key), config[key])
+}
+
+func parseNode(node interface{}) (Targeting, error) {
+	switch n := node.(type) {
+	case []interface{}:
+	case map[interface{}]interface{}:
+		key := firstKey(n)
+		return mapOperatorNode(strings.ToLower(key), n[key])
+	case map[string]interface{}:
+		key := firstKeyStr(n)
+		return mapOperatorNode(strings.ToLower(key), n[key])
+	default:
+		panic(n)
```

Comment:
should this just return an error rather than panicking?

Reply:
Yes, never panic. Still got some cruft to clean up.